File: services/customer.service.js
/**
 * BarberEase – Customer Service
 * 
 * Business logic for customer profile management.
 */

const bcrypt = require('bcrypt');
const config = require('../config/app');
const { Customer } = require('../models');

/** Get customer profile by ID. */
async function getProfile(customerId) {
  const customer = await Customer.findByPk(customerId);
  if (!customer) {
    const error = new Error('Customer not found');
    error.statusCode = 404;
    throw error;
  }
  return customer;
}

/** Update customer profile. */
async function updateProfile(customerId, data) {
  const customer = await Customer.findByPk(customerId);
  if (!customer) {
    const error = new Error('Customer not found');
    error.statusCode = 404;
    throw error;
  }

  const updateData = {};
  if (data.name) updateData.name = data.name;
  if (data.phone) updateData.phone = data.phone;
  if (data.address !== undefined) updateData.address = data.address;
  if (data.gender) updateData.gender = data.gender;

  await customer.update(updateData);
  return customer;
}

/** Change customer password. */
async function changePassword(customerId, currentPassword, newPassword) {
  const customer = await Customer.scope('withPassword').findByPk(customerId);
  if (!customer) {
    const error = new Error('Customer not found');
    error.statusCode = 404;
    throw error;
  }

  const isMatch = await bcrypt.compare(currentPassword, customer.password);
  if (!isMatch) {
    const error = new Error('Current password is incorrect');
    error.statusCode = 400;
    throw error;
  }

  const hashedPassword = await bcrypt.hash(newPassword, config.bcrypt.saltRounds);
  await customer.update({ password: hashedPassword });

  return { message: 'Password changed successfully' };
}

/** Get all customers (admin). */
async function getAllCustomers() {
  return await Customer.findAll({
    order: [['created_at', 'DESC']]
  });
}

/** Toggle customer active status (admin). */
async function toggleCustomerStatus(customerId) {
  const customer = await Customer.findByPk(customerId);
  if (!customer) {
    const error = new Error('Customer not found');
    error.statusCode = 404;
    throw error;
  }

  await customer.update({ is_active: !customer.is_active });
  return customer;
}

/** Top up customer wallet balance. */
async function topUpWallet(customerId, amount) {
  const customer = await Customer.findByPk(customerId);
  if (!customer) {
    const error = new Error('Customer not found');
    error.statusCode = 404;
    throw error;
  }

  const currentBalance = parseFloat(customer.wallet_balance || 0);
  const topUpAmount = parseFloat(amount || 0);

  if (isNaN(topUpAmount) || topUpAmount <= 0) {
    const error = new Error('Please provide a valid top up amount greater than zero.');
    error.statusCode = 400;
    throw error;
  }

  await customer.update({
    wallet_balance: currentBalance + topUpAmount
  });

  return customer;
}

module.exports = {
  getProfile,
  updateProfile,
  changePassword,
  getAllCustomers,
  toggleCustomerStatus,
  topUpWallet
};

File: src/main/java/com/barberease/services/CustomerService.java
package com.barberease.services;

import com.barberease.models.Customer;
import com.barberease.repositories.CustomerRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;

@Service
@Transactional
public class CustomerService {

    @Autowired
    private CustomerRepository customerRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    public Customer getProfile(Long customerId) {
        return customerRepository.findById(customerId)
                .orElseThrow(() -> new IllegalArgumentException("Customer not found"));
    }

    public Customer updateProfile(Long customerId, Customer data) {
        Customer customer = getProfile(customerId);

        if (data.getName() != null) customer.setName(data.getName());
        if (data.getPhone() != null) customer.setPhone(data.getPhone());
        if (data.getAddress() != null) customer.setAddress(data.getAddress());
        if (data.getGender() != null) customer.setGender(data.getGender());

        return customerRepository.save(customer);
    }

    public void changePassword(Long customerId, String currentPassword, String newPassword) {
        Customer customer = getProfile(customerId);

        if (!passwordEncoder.matches(currentPassword, customer.getPassword())) {
            throw new IllegalArgumentException("Current password is incorrect");
        }

        customer.setPassword(passwordEncoder.encode(newPassword));
        customerRepository.save(customer);
    }

    public List<Customer> getAllCustomers() {
        return customerRepository.findAll();
    }

    public Customer toggleCustomerStatus(Long customerId) {
        Customer customer = getProfile(customerId);
        customer.setActive(!customer.isActive());
        return customerRepository.save(customer);
    }

    public Customer topUpWallet(Long customerId, BigDecimal amount) {
        Customer customer = getProfile(customerId);

        if (amount == null || amount.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Please provide a valid top up amount greater than zero.");
        }

        customer.setWalletBalance(customer.getWalletBalance().add(amount));
        return customerRepository.save(customer);
    }
}
